src/config/voice.c

```diff
@@ -28,55 +30,54 @@
 const char SECTION_VOICE_GLOBAL[] = "global";
 const char SECTION_VOICE_CUSTOM[] = "custom";
 
-static const char* getfield(char* value, int num) {
-    const char* tok;
-    for (tok = strtok(value, ","); tok && *tok; tok = strtok(NULL, ",\n")) {
-        if (!--num)
-            return tok;
-    }
-    return "";
-}
-
 static int ini_handler(void* user, const char* section, const char* name, const char* value)
 {
-    (void) user;
-    char tmp[100];
-    strlcpy(tmp, value, 100);
-    int duration = atoi(getfield(tmp,2));
-#if HAS_MUSIC_CONFIG
-    char label[MAX_VOICE_LABEL];
-    strlcpy(label,getfield(tmp, 1), MAX_VOICE_LABEL);
-#endif
+    u16 req_id = *((long*)user);
+    u16 id = atoi(name);
+    const char* ptr = value;
+    u8 k = 0;
+    u16 duration = 0;
+
+    while (*ptr && *ptr != ',') {
+        ptr++;
+        k++;
+    }
+    if (*ptr == ',') {
+        duration = atoi(ptr + 1);
+        }
 
-    if (MATCH_SECTION(SECTION_VOICE_GLOBAL)) {
-        for (int i = 0; i < CUSTOM_ALARM_ID; i++) {
-            snprintf(tempstring, 4, "%d", i);
-            if (MATCH_KEY(tempstring)) {
-                voice_map[i].duration = duration;
-                voice_map[i].id = i;
 #if HAS_MUSIC_CONFIG
-                strcpy(voice_map[i].label, label);
+    if ( k && (req_id != MAX_VOICEMAP_ENTRIES) && (req_id == id) ) {
+        strlcpy(tempstring, value, k+1);
```

Comment:
still implict return tempstring?

Reply:
I know, but ini handler can only return into. How else to solve this problem? If we don't rely on ini.c code, rom size will likely grow quite a bit...

Reply:
Can u add some comment to make the logic explicit? I am fine with that.